BusinessAdapter.FP/Receiving/FpMessageReceiver.cs

```diff
@@ -0,0 +1,301 @@
+            FpInboxMessage[] availableMessages,
+            int messageLimit,
+            ReceiveStatus receiveStatus)
+        {
+            var messagesForRetry = new List<FpInboxMessage>();
+            var exceptions = new List<Exception>();
+
+            for (var i = 0; i < messageLimit; i++)
+            {
+                try
+                {
+                    await ProcessSingleMessageAsync(receiveContext, availableMessages[i], receiveStatus);
+                }
+                catch (Exception ex)
+                {
+                    receiveStatus.AddFailedReceivedMessage(availableMessages[i], ex);
+                    messagesForRetry.Add(availableMessages[i]);
+                    exceptions.Add(new RetryableException(
+                        $"Error while receiving message with identification {availableMessages[i].MessageId}.", ex));
+                }
+            }
+
+            if (messagesForRetry.Count != 0)
+            {
+                throw new AggregateException(
+                    "At least one error occurred. Details can be found in the inner exceptions.", exceptions);
+            }
+        }
+
+        private async Task ProcessSingleMessageAsync(
+            KeyValuePair<MessageReceiveInfo, IBusinessApiGateway> receiveContext, FpInboxMessage message,
+            ReceiveStatus receiveStatus)
+        {
+            var result = await receiveContext.Value.ReceiveMessageAsync(message);
+
+            if (!result.WasSuccessful)
+            {
+                if (HandleTooManyRequestError(result.ResponseStatusCode!.Value))
+                {
+                    receiveContext.Key.HasTooManyRequestsError = true;
+                    return;
+                }
+
+                throw new InvalidOperationException("Error while receiving AS4 messages.");
+            }
+
+            var fileName = fpFileRepo.StoreXmlFileTo(result.Message, receiveOptions.Directory);
+
+            var ackResponse = await receiveContext.Value.AcknowledgeReceivedMessageAsync(result.Message);
+            if (!ackResponse.WasSuccessful)
+            {
+                fpFileRepo.DeleteFile(fileName);
+
+                if (HandleTooManyRequestError(ackResponse.ResponseStatusCode!.Value))
+                {
+                    receiveContext.Key.HasTooManyRequestsError = true;
+                    return;
+                }
+
+                throw new InvalidOperationException("Error while acknowledging AS4 messages.");
+            }
+
+            receiveStatus.AddSuccessfulReceivedMessage(message);
+        }
+
+        private void LogFinalStatus(ReceiveStatus receiveStatus,
+            Dictionary<MessageReceiveInfo, IBusinessApiGateway> as4BusinessApiClients,
+            List<string> marketPartnersWithoutCertificate)
+        {
+            foreach (var as4BusinessApiClient in as4BusinessApiClients)
+            {
+                as4BusinessApiClient.Value.Dispose();
```

Comment:
wrong place - assuming we remove the logging, then we also remove the disposing?